front/lib/api/llm/clients/fireworks/utils/conversation_to_openai.ts

```diff
@@ -0,0 +1,154 @@
+  switch (content.type) {
+    case "text":
+      return content;
+    case "image_url":
+      return {
+        type: "image_url",
+        image_url: {
+          url: content.image_url.url,
+        },
+      };
+    case "text_content":
+      return content.value;
+    case "reasoning":
+      // OpenAI doesn't have a specific reasoning type, so we treat it as text
+      // The reasoning will be part of the assistant's response
+      assert(content.value.reasoning, "Reasoning content is missing reasoning");
+      return content.value.reasoning;
+    default:
+      assertNever(content);
+  }
+}
+
+function toAssistantMessage(
+  message:
+    | AssistantFunctionCallMessageTypeModel
+    | AssistantContentMessageTypeModel
+): ChatCompletionAssistantMessageParam {
+  assert(message.contents, "Assistant message is missing contents");
+
+  const textContents = message.contents
+    .filter(
+      (c): c is TextContentType | ReasoningContentType =>
+        c.type !== "function_call"
+    )
+    .map(toContentPart)
+    .filter((c): c is string => typeof c === "string");
+
+  const toolCalls = message.contents
+    .filter((c): c is FunctionCallContentType => c.type === "function_call")
+    .map((fc) => ({
+      id: fc.value.id,
+      type: "function" as const,
+      function: {
+        name: fc.value.name,
+        arguments: fc.value.arguments,
+      },
+    }));
+
+  const result: ChatCompletionAssistantMessageParam = {
+    role: "assistant",
+  };
+
+  if (textContents.length > 0) {
+    result.content = textContents.join("");
+  }
+
+  if (toolCalls.length > 0) {
+    result.tool_calls = toolCalls;
+  }
+
+  return result;
+}
+
+export function toTool(
+  specification: AgentActionSpecification
+): ChatCompletionTool {
+  const properties = specification.inputSchema.properties ?? {};
+
+  return {
+    type: "function",
+    function: {
+      name: specification.name,
+      description: specification.description,
+      parameters: {
+        type: "object",
+        properties,
+        required: Object.keys(properties),
+        additionalProperties: false,
+      },
+      strict: false,
+    },
+  };
+}
+
+export function toMessage(
+  message: ModelMessageTypeMultiActionsWithoutContentFragment
+): ChatCompletionMessageParam {
+  switch (message.role) {
+    case "user": {
+      return {
+        role: "user",
+        content: message.content.map(toContentPart),
+      };
+    }
+    case "function": {
+      const content =
+        typeof message.content === "string"
+          ? message.content
+          : message.content
+              .map((c) => {
+                const part = toContentPart(c);
+                // If it's a string, return it directly
+                if (typeof part === "string") {
+                  return part;
+                }
+                // If it's a text content part, extract the text
+                if ("text" in part) {
+                  return part.text;
+                }
```

Comment:
Looking at the implementation of `toContentPart`, I think we never return `ChatCompletionContentPartText`